draft-patton-cfrg-vdaf.md

```diff
@@ -705,6 +705,159 @@ TODO
+been evaluated yet for `state`, returning a secret-shared value.
+
+### Malicious sketching for IDPFs
+
+~~~
+def verify_sketch_round_1(dpf_output: Vec[Value], correlation_share, joint_randomness: Bytes, level: Unsigned):
+  # Expand common randomness and compute verification message
+  r = expand_rand(joint_randomness, len(dpf_output), level)
+  r_squared = [v*v for v in r]
+  ver_msg_1 = inner_product(
+    [value[0] for value in dpf_values,
+    r
+  )
+  ver_msg_2 = inner_product(
+    [value[0] for value in dpf_values,
+    r_squared
+  )
+  ver_msg_3 = inner_product(
+    [value[1] for value in dpf_values],
+    r
+  )
+  outbound_message = {
+    ver_msg_1: ver_msg_1 + correlation_share.a
+    ver_msg_2: ver_msg_2 + correlation_share.b
+    ver_msg_3: ver_msg_3 + correlation_share.c
+  }
+  state = {
+    ver_msg_1: ver_msg_1,
+    ver_msg_2: ver_msg_2,
+    ver_msg_3: ver_msg_3,
+
+  return (state, outbound_message)
+~~~
+
+~~~
+def verify_sketch_round_2(state, inbound_message, correlation_share):
+  ver_msg_1 = state.ver_msg_1 + inbound_message.ver_msg_1
+  ver_msg_2 = state.ver_msg_2 + inbound_message.ver_msg_2
+  ver_msg_3 = state.ver_msg_3 + inbound_message.ver_msg_3
+  verification = ver_msg_1**2 - ver_msg_2 - ver_msg_3 + \
+    correlation_share.A * ver_msg_1 + correlation_share.B
+  if verification == 0:
+    return True
+  else:
+    return False
+~~~
+
+## Construction
+
+~~~
+def vdaf_setup(n: Unsigned):
+  public_param = {
+    n: n
+  }
+  return (public_param, None)
+~~~
+
+~~~
+def vdaf_input(public_param, input):
+  assert(log2(input) < public_param.n)
+  beta = []
+  correlation_shares_0 = [], correlation_shares_1 = []
+  for l in range(public_param.n):
+    # Construct values of the form (1, k), where k is a random field element.
+    _, field = get_value_type(l)
+    k = get_random_element(field)
+    beta += [(1, k)]
+    # Create secret shares of correlations to aid the servers' computation
+    a, b, c = get_random_element(field), get_random_element(field), \
+      get_random_element(field)
+    A = -2*a+k
+    B = a*a + b - a * k + c
+    correlation_shares_0 += [{
+      a: get_random_element(field),
+      b: get_random_element(field),
+      c: get_random_element(field),
+      A: get_random_element(field),
+      B: get_random_element(field),
+    }]
+    correlation_shares_1 += [{
+      a: a - correlation_shares_0[l]
+      b: b - correlation_shares_0[l]
+      c: c - correlation_shares_0[l]
+      A: A - correlation_shares_0[l]
+      B: B - correlation_shares_0[l]
+    }]
+  key_0, key_1 = idpf_gen(input, beta)
+  return ((key_0, correlation_shares_0), (key_1, correlation_shares_1))
+~~~
+
+~~~
+def vdaf_start(verify_param, public_param, nonce, input_share):
+  key, correlation_share = input_share
+
+  # Expand left and right children of the DPF tree.
+  state_left, value_left = idpf_eval_next(None, key, 0)
+  state_right, value_right = idpf_eval_next(None, key, 1)
```

Comment:
I'm having a hard time understanding what's going on here. It seeems like you want the IDPF api to be stateful, but the VDAF api is not. So how does the state of the IPDF get passed through to subsequent VDAF evaluations?

Unless I'm misunderstanding, I think the answer is that, at least as far as this spec is concerned, the IDPF api needs to be stateless.